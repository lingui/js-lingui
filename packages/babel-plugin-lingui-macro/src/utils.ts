export const makeCounter =
  (index = 0) =>
  () =>
    index++
