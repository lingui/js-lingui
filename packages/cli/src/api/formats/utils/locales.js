// @flow
import plurals from "make-plural"

export type LocaleObject = {
  language: string,
  country?: string
}

/**
 * Groups:
 * 0. and 1. locale code (1. is useful, when RegExp is part of larger pattern)
 * 2. language code
 * 4. country code
 * @type {RegExp}
 */
export const localeRe = new RegExp(/(([a-z]+)([_-]([a-zA-Z]+))?)(\\|\/)?/)

export function isValid(locale: string): boolean {
  const match = localeRe.exec(locale)
  return (
    match &&
    match[0] === locale && // locale has valid format
    match[2] in plurals // language is valid (we have plurals for it)
  )
}

export function parse(locale: string): ?LocaleObject {
  const match = localeRe.exec(locale)
  if (!match || match[0] !== locale) return null

  return {
    language: match[2],
    country: match[4]
  }
}
