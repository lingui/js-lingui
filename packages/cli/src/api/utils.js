import fs from "fs"
import path from "path"
import chalk from "chalk"
import { score } from "fuzzaldrin"

export function removeDirectory(dir, keep = false) {
  if (!fs.existsSync(dir)) return
  const list = fs.readdirSync(dir)

  for (let i = 0; i < list.length; i++) {
    const filename = path.join(dir, list[i])
    const stat = fs.statSync(filename)

    if (filename === "." || filename === "..") {
      // pass these files
    } else if (stat.isDirectory()) {
      // rmdir recursively
      removeDirectory(filename)
    } else {
      fs.unlinkSync(filename)
    }
  }

  if (!keep) {
    fs.rmdirSync(dir)
  }
}

export function prettyOrigin(origins) {
  try {
    return origins.map(origin => origin.join(":")).join(", ")
  } catch (e) {
    return ""
  }
}

/**
 * .. js:function:: helpMisspelledCommand(command [, availableCommands = []])
 *    :param: command - command passed to CLI
 *    :param: availableCommands - all commands defined in commander.js
 *
 *    If unknown commands is passed to CLI, check it agains all available commands
 *    for possible misspelled letter. Output help with suggestions to console.
 */
export function helpMisspelledCommand(command, availableCommands = []) {
  const commandNames = availableCommands.map(command => command.name())

  // if no command is supplied, then commander.js shows help automatically
  if (!command || commandNames.includes(command)) {
    return
  }

  const suggestions = commandNames
    .map(name => ({
      name,
      score: score(name, command.slice(0, name.length))
    }))
    .filter(nameScore => nameScore.score > 0)
    .slice(0, 3)
    .map(commandScore => chalk.inverse(commandScore.name))
    .join(", ")

  console.log(
    `lingui: command ${command} is not a lingui command. ` +
      `See 'lingui --help' for the list of available commands.`
  )

  if (suggestions) {
    console.log()
    console.log(`Did you mean: ${suggestions}?`)
  }
}

export const splitOrigin = origin => origin.split(":")

export const joinOrigin = origin => origin.join(":")
