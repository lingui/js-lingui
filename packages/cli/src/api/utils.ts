import fs from "fs"
import path from "path"
import normalize from "normalize-path"

export const PATHSEP = "/" // force posix everywhere

export function prettyOrigin(origins: [filename: string, line?: number][]) {
  try {
    return origins.map((origin) => origin.join(":")).join(", ")
  } catch (e) {
    return ""
  }
}

export function replacePlaceholders(
  input: string,
  values: Record<string, string>
): string {
  return input.replace(/\{([^}]+)}/g, (m, placeholder) => {
    return values[placeholder] ?? m
  })
}

export const splitOrigin = (origin: string) => {
  const [file, line] = origin.split(":")
  return [file, line ? Number(line) : null] as [file: string, line: number]
}

export const joinOrigin = (origin: [file: string, line?: number]): string =>
  origin.join(":")

export async function readFile(fileName: string): Promise<string | undefined> {
  try {
    return (await fs.promises.readFile(fileName)).toString()
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code != "ENOENT") {
      throw err
    }
  }
}

async function mkdirp(dir: string): Promise<void> {
  try {
    await fs.promises.mkdir(dir, {
      recursive: true,
    })
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code != "EEXIST") {
      throw err
    }
  }
}

export function isDirectory(filePath: string) {
  try {
    return fs.lstatSync(filePath).isDirectory()
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code != "ENOENT") {
      throw err
    }
  }
}

export async function writeFile(
  fileName: string,
  content: string
): Promise<void> {
  await mkdirp(path.dirname(fileName))
  await fs.promises.writeFile(fileName, content)
}

export async function writeFileIfChanged(
  filename: string,
  newContent: string
): Promise<void> {
  const raw = await readFile(filename)

  if (raw) {
    if (newContent !== raw) {
      await writeFile(filename, newContent)
    }
  } else {
    await writeFile(filename, newContent)
  }
}

export function hasYarn() {
  return fs.existsSync(path.resolve("yarn.lock"))
}

export function makeInstall(packageName: string, dev: boolean = false) {
  const withYarn = hasYarn()

  return withYarn
    ? `yarn add ${dev ? "--dev " : ""}${packageName}`
    : `npm install ${dev ? "--save-dev" : "--save"} ${packageName}`
}

/**
 * Normalize Windows backslashes in path so they look always as posix
 */
export function normalizeSlashes(path: string) {
  return path.replace("\\", "/")
}

/**
 * Remove ./ at the beginning: ./relative  => relative
 *                             relative    => relative
 * Preserve directories:       ./relative/ => relative/
 * Preserve absolute paths:    /absolute/path => /absolute/path
 */
export function normalizeRelativePath(sourcePath: string): string {
  if (path.isAbsolute(sourcePath)) {
    // absolute path
    return normalize(sourcePath, false)
  }

  // https://github.com/lingui/js-lingui/issues/809
  const isDir = isDirectory(sourcePath)

  return (
    normalize(path.relative(process.cwd(), sourcePath), false) +
    (isDir ? "/" : "")
  )
}
