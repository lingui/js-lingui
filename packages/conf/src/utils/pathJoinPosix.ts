import path from "path"

// Enforce posix path delimiters internally
export const pathJoinPosix = (...values: string[]) =>
  path
    // normalize double slashes
    .join(...values)
    // convert platform specific path.sep to posix
    .split(path.sep)
    .join("/")
