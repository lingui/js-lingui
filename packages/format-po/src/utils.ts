export function normalizePlaceholderValue(text: string) {
  return text.replace(/\n/g, " ").replace(/\s{2,}/g, " ")
}
