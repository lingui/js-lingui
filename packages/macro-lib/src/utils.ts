import type { NodePath } from "@babel/traverse"
import type { ImportDeclaration, Program } from "@babel/types"
import type * as babelTypes from "@babel/types"

export const makeCounter =
  (index = 0) =>
  () =>
    index++

/**
 * Filtering nested macro calls
 *
 * <Macro>
 *   <Macro /> <-- this would be filtered out
 * </Macro>
 */
export function isRootPath(allPath: NodePath[]) {
  return (node: NodePath) =>
    (function traverse(path): boolean {
      if (!path.parentPath) {
        return true
      } else {
        return !allPath.includes(path.parentPath) && traverse(path.parentPath)
      }
    })(node)
}

export function addImport(
  t: typeof babelTypes,
  program: Program,
  module: string,
  importName: string
) {
  const linguiImport = program.body.find(
    (importNode) =>
      t.isImportDeclaration(importNode) &&
      importNode.source.value === module &&
      // https://github.com/lingui/js-lingui/issues/777
      importNode.importKind !== "type"
  ) as ImportDeclaration

  const tIdentifier = t.identifier(importName)
  // Handle adding the import or altering the existing import
  if (linguiImport) {
    if (
      linguiImport.specifiers.findIndex(
        (specifier) =>
          t.isImportSpecifier(specifier) &&
          t.isIdentifier(specifier.imported, { name: importName })
      ) === -1
    ) {
      linguiImport.specifiers.push(t.importSpecifier(tIdentifier, tIdentifier))
    }
  } else {
    program.body.unshift(
      t.importDeclaration(
        [t.importSpecifier(tIdentifier, tIdentifier)],
        t.stringLiteral(module)
      )
    )
  }
}

export function throwNonMacroContextError(packageName: string) {
  throw new Error(
    `The macro you imported from "${packageName}" is being executed outside the context of compilation with babel-plugin-macros. ` +
      `This indicates that you don't have the babel plugin "babel-plugin-macros" or "@lingui/swc-plugin" configured correctly.` +
      `\n Please see the documentation for how to configure babel-plugin-macros properly: ` +
      "https://github.com/kentcdodds/babel-plugin-macros/blob/main/other/docs/user.md" +
      `\n For SWC Version: https://lingui.dev/ref/swc-plugin`
  )
}
