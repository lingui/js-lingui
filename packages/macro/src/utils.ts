import * as R from "ramda"

/**
 * Custom zip method which takes length of the larger array
 * (usually zip functions use the `smaller` length, discarding values in larger array)
 */
export function zip<A, B>(a: A[], b: B[]): [A, B][] {
  return R.range(0, Math.max(a.length, b.length)).map((index) => [
    a[index],
    b[index],
  ])
}

export const makeCounter = (index = 0) => () => index++
