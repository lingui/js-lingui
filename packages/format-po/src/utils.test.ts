import { normalizePlaceholderValue } from "./utils"

describe("normalizePlaceholderValue", () => {
  it.each([
    `user 
    ? user.name 
    : null`,
    "userName",
  ])("Should normalize whitespaces", (input) => {
    expect(normalizePlaceholderValue(input)).toMatchSnapshot()
  })
})
