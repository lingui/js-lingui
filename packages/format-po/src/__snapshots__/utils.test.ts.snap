// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizePlaceholderValue Should normalize whitespaces 1`] = `user ? user.name : null`;

exports[`normalizePlaceholderValue Should normalize whitespaces 2`] = `userName`;
